bindings/python/pymongocrypt/asynchronous/mongocrypt.py

```diff
@@ -0,0 +1,682 @@
+
+    def rewrap_many_data_key_context(self, filter, provider, master_key):
+        """Creates a context to use for rewrapping many data keys.
+
+        :Parameters:
+          - `filter`: A document used to filter the data keys.
+          - `provider`: (optional) The name of a different kms provider.
+          - `master_key`: Optional document for the given provider.
+            MUST have the fields corresponding to the
+            given provider as specified in master_key. master_key MUST NOT be
+            given if it is not applicable for the given provider.
+
+        :Returns:
+          A :class:`RewrapManyDataKeyContext`.
+        """
+        return RewrapManyDataKeyContext(self._create_context(), self.__opts.kms_providers, filter, provider, master_key, self.__callback)
+
+
+class MongoCryptContext(object):
+    __slots__ = ("__ctx", "__kms_providers")
+
+    def __init__(self, ctx, kms_providers):
+        """Abstracts libmongocrypt's mongocrypt_ctx_t type.
+
+        :Parameters:
+          - `ctx`: A mongocrypt_ctx_t. This MongoCryptContext takes ownership
+            of the underlying mongocrypt_ctx_t.
+          - `kms_providers`: The KMS provider map.
+        """
+        self.__ctx = ctx
+        self.__kms_providers = kms_providers
+
+    def _close(self):
+        """Cleanup resources."""
+        if self.__ctx is None:
+            return
+        lib.mongocrypt_ctx_destroy(self.__ctx)
+        self.__ctx = None
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._close()
+
+    @property
+    def state(self):
+        """The current state of the mongocrypt_ctx_t."""
+        return lib.mongocrypt_ctx_state(self.__ctx)
+
+    def _raise_from_status(self):
+        status = lib.mongocrypt_status_new()
+        try:
+            lib.mongocrypt_ctx_status(self.__ctx, status)
+            exc = MongoCryptError.from_status(status)
+        finally:
+            lib.mongocrypt_status_destroy(status)
+        raise exc
+
+    def mongo_operation(self):
+        """Returns the mongo operation to execute as bson bytes."""
+        with MongoCryptBinaryOut() as binary:
+            if not lib.mongocrypt_ctx_mongo_op(self.__ctx, binary.bin):
+                self._raise_from_status()
+            return binary.to_bytes()
+
+    def add_mongo_operation_result(self, document):
+        """Adds the mongo operation's command response.
+
+        :Parameters:
+          - `document`: A raw BSON command response document.
+        """
+        with MongoCryptBinaryIn(document) as binary:
+            if not lib.mongocrypt_ctx_mongo_feed(self.__ctx, binary.bin):
+                self._raise_from_status()
+
+    def complete_mongo_operation(self):
+        """Completes the mongo operation."""
+        if not lib.mongocrypt_ctx_mongo_done(self.__ctx):
+            self._raise_from_status()
+
+    async def ask_for_kms_credentials(self):
```

Comment:
I think we should make kms_providers a property, hoist this whole file up to the top level, and call `_ask_for_kms_credentials(ctx.kms_providers)`.

Reply:
A property? Can you explain what you mean? `_ask_for_kms_credentials` does network IO so it needs to have both async and sync versions.

Reply:
The only async aspect of this file is this method.  Instead, I'm suggesting we have `MongoCryptContext.kms_providers` and then instead of calling `ctx.ask_for_kms_credentials`, we call the appropriate `_ask_for_kms_credentials` with `ctx.kms_providers`.  Only the `_ask_for_kms_credentials` function needs to be async/syncified.

Reply:
Ahh makes sense. The only other part that differs is this line: 

```
        if not isinstance(callback, AsyncMongoCryptCallback):
            raise TypeError("callback must be a AsyncMongoCryptCallback")
```

If we're fine with that checking for both sync and async callbacks, we can do this change.

Reply:
Yeah I think that's fine.